### DISCUSSION TITLE
Regarding the addition of printers (devices) to the Papple-based printer application using IPP calls.

### DISCUSSION THREAD

**Discussion by Kaushik1216** [michaelrsweet/pappl#372] [Q&A]
Hello Everyone!

I'm working on a GSoC '24 project, with one of the main goals being to add printers (devices) to the printer application through IPP calls.

Steps:

Using device ID information, I make an IPP call to the printer application to retrieve the appropriate driver.
Using the driver and additional information, I then make a call to the printer application to add the printer.
Below are the implementations of both IPP calls.

I believe I might be using incorrect options for the IPP calls. Could anyone help me troubleshoot this? I have debugged the code but am not getting meaningful error messages.

code

```c
gboolean
cph_cups_printer_add (CphCups    *cups,
                      const char *printer_name,
                      const char *printer_uri,
                      const char *ppd_file,
                      const char *info,
                      const char *location)
{

        ipp_t		        *request,		
		                *response;		
        ipp_attribute_t         *attr;		
        http_t	                *http;			
        int                     timeout_param = CUPS_TIMEOUT_DEFAULT;

        http = httpConnect2("kaushik.local", 8001, NULL, AF_UNSPEC,
                           HTTP_ENCRYPTION_IF_REQUESTED, 1, 30000, NULL);

        request = ippNewRequest(IPP_OP_PAPPL_FIND_DRIVERS);
        ippAddString(request, IPP_TAG_OPERATION, IPP_CONST_TAG(IPP_TAG_URI), "system-uri", NULL, "ipp://localhost/ipp/system");
        // I will pass here correct device id for devices
        //ippAddString(request, IPP_TAG_OPERATION, IPP_CONST_TAG(IPP_TAG_TEXT), "smi55357-device-id", NULL, "MFG:HP;CMD:PJL,PML,DW-PCL;MDL:HP LaserJet Tank MFP 260x;CLS:PRINTER;DES:HP LaserJet Tank MFP 2606dn;MEM:MEM=308MB;PRN:381U0A;S:0300000000000000000000000000000000;COMMENT:RES=600x2;LEDMDIS:USB#ff#04#01;CID:HPLJPCLMSMV2;MCT:MF;MCL:FL;MCV:4.2;");
        response = cupsDoRequest(http, request, "/ipp/system");

        attr = ippFindAttribute(response, "smi55357-driver-col", IPP_TAG_BEGIN_COLLECTION);
        const char *driver = NULL;

        if (attr != NULL) {
            driver = ippGetString(attr, 0, NULL);
        }

        request = ippNewRequest(IPP_OP_CREATE_PRINTER);

        ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_URI, "system-uri", NULL, "ipp://localhost/ipp/system");
        ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_KEYWORD, "printer-service-type", NULL, "print");
        ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_KEYWORD, "smi55357-driver", NULL, driver);
        ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_URI, "smi55357-device-uri", NULL, printer_uri);
        ippAddString(request, IPP_TAG_PRINTER, IPP_TAG_NAME, "printer-name", NULL, printer_name);

        response = cupsDoRequest(http, request, "/ipp/system");

        gboolean status = FALSE;
        if (cupsLastError() != IPP_STATUS_OK) {
                status = false;
        }
        ippDelete(response);
        httpClose(http);

        return status;
 }
```

**Comment by michaelrsweet** [comment]
It would be helpful to know what errors you are getting, meaningful or not. Calling `cupsLastErrorString()` will at least give you the human-readable error message from the printer application, as will looking at the logs from the printer application...

One thing of note in the included code - connecting to the mDNS hostname will almost certainly result in a remote network connection (even on the same machine) which will fail since remote admin is disabled by default in printer applications (for what should be obvious security reasons...)

**Comment by Kaushik1216** [comment]
First of all, sorry for the late reply!
Is there any other way to connect with the Printer Application and add printers without directly interacting with the PrinterApp application?

**Reply by michaelrsweet** [reply]
No, the printers are managed by the Printer Application so you have to talk to the printer application to add printers.

**Reply by Kaushik1216** [reply]
Yes, I have written code to connect to the printer application and add a printer. The code is attached in my first comment. I will also try to capture any errors.

Could you please check if my code is correct?

**Reply by michaelrsweet** [reply]
As I noted before, using the mDNS hostname is problematic, and you are hardcoding the system-uri value (which should include the proper scheme, hostname, and port number).

The "smi55357-driver-col" attribute is a 1setOf collection, so you can't use ippGetString with it. You need to look up the "smi55357-driver" member attribute, e.g.:

```c
driver = ippGetString(ippFindAttribute(ippGetCollection(attr, 0), "smi55357-driver", IPP_TAG_KEYWORD), 0, NULL);
```
